src/Framework/FileClassifier.cs

```diff
@@ -0,0 +1,250 @@
+                {
+                    if (count < 1)
+                        return path;
+
+                    var parent = Directory.GetParent(path);
+
+                    while (count > 1 && parent?.Parent != null)
+                    {
+                        parent = parent.Parent;
+                        count--;
+                    }
+
+                    return parent?.FullName ?? path;
+                }
+            }
+#endif
+        }
+
+        /// <summary>
+        ///     Shared singleton instance.
+        /// </summary>
+        public static FileClassifier Shared => s_sharedInstance.Value;
+
+        /// <summary>
+        ///     Try add paths found into set of known immutable paths.
+        ///     Project files under any of these folders are considered non-modifiable.
+        /// </summary>
+        /// <remarks>
+        ///     This value is used by <see cref="IsNonModifiable" />.
+        ///     Files in the NuGet package cache are not expected to change over time, once they are created.
+        /// </remarks>
+        /// <remarks>
+        ///     Example value: <c>"C:\Users\myusername\.nuget\;D:\LocalNuGetCache\"</c>
+        /// </remarks>
+        public void RegisterImmutableDirectories(string? nuGetPackageFolders)
+        {
+            if (nuGetPackageFolders?.Length > 0)
+            {
+                string[] folders = nuGetPackageFolders.Split(s_semicolonDelimiter, StringSplitOptions.RemoveEmptyEntries);
+                foreach (string folder in folders)
+                {
+                    RegisterImmutableDirectory(folder);
+                }
+            }
+        }
+
+        private void RegisterImmutableDirectory(string? directory)
+        {
+            if (directory?.Length > 0)
+            {
+                string d = EnsureTrailingSlash(directory);
+
+                if (_knownImmutableDirectories.TryAdd(d, d))
+                {
+                    _knownImmutableDirectoriesSnapshot = new List<string>(_knownImmutableDirectories.Values);
+                }
+            }
+        }
+
+        private static string EnsureTrailingSlash(string fileSpec)
+        {
+            if (fileSpec.Length >= 1)
+            {
+                char lastChar = fileSpec[fileSpec.Length - 1];
+                if (lastChar != Path.DirectorySeparatorChar && lastChar != Path.AltDirectorySeparatorChar)
+                {
+                    fileSpec += Path.DirectorySeparatorChar;
+                }
+            }
+
+            return fileSpec;
+        }
+
+        /// <summary>
+        ///     Gets whether a file is expected to not be modified in place on disk once it has been created.
+        /// </summary>
+        /// <param name="filePath">The path to the file to test.</param>
+        /// <returns><see langword="true" /> if the file is non-modifiable, otherwise <see langword="false" />.</returns>
+        public bool IsNonModifiable(string filePath)
```

Comment:
My English skill is not as strong as to understand if `NonModifiable` is better than `Immutable` or `Unchangable` so  I stickled to whatever was used in source I have adopted it from [CPS](https://github.com/dotnet/project-system/blob/6dcc31fa39ad6c922bfd73af553e37e7cdfd8b80/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectFileClassifier.cs).
Also because the default if is mostly like `if (IsNonModifiable) {check_and_skip_check}` and having it `if (!IsModifiable) {check_and_skip_check}` would not make it any less readable.

Reply:
I'd prefer IsFixed...but it's also reasonable to align with CPS. The bigger issue is if you get `!IsNonModifiable` at some point, at which point you have to do mental gymnastics to figure out what that actually means.